### PR TITLE
Allow setting custom cipher strings in the openssl config file.

### PR DESCRIPTION
**Pull request by nmav**
When the PROFILE=KEYWORD is specified as a cipher string, openssl will
expand the keyword to the defined cipher string in the configuration
file under section "cipher_profiles". That requires the default
(or any other) configuration file to be loaded.

An example entry in the configuration file:
[ cipher_profiles ]
PROFILE1 = ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384
PROFILE2 = HIGH

Also in: http://rt.openssl.org/Ticket/Display.html?id=3299

### CODE REVIEW

File: ssl/ssl_ciph.c [hunk 1]
```diff
@@ -1479,6 +1549,7 @@ STACK_OF(SSL_CIPHER) *ssl_create_cipher_list(const SSL_METHOD *ssl_method, STACK
      */
     if (!ssl_cipher_strength_sort(&head, &tail)) {
         OPENSSL_free(co_list);
+	OPENSSL_free(new_rules);
```
Comment (frankmorgner) [inline ssl/ssl_ciph.c]: bad indenting

**Comment by richsalz** [comment]
+1

**Comment by nmav** [comment]
@richsalz Is there some time plan for that?
We would really like to enhance the ciphersuite strings of openssl for one to be able to specify signature algorithms, protocols and other parameters; and that patch is the first step to that. Is that also something you would be interested as well? What we would like to achieve is have a central location for default settings in an OS rather than every application setting its own.

Another issue we figured is that application that use chroot() may not benefit from this approach. Is that something you would be interested to addressing with a later patch? (my first impression is that config file load could move to a constructor).

**Comment by richsalz** [comment]
No, there is no timetable ☹  I don’t see the point of breaking up the c iphersuite strings into finer level of detail as it’s more likely to be a way for users to get confused and have bad configurations (since TLS itself doesn’t let you do that), but maybe I’m wrong.

I agree with centralized settings, which is why I liked this.  Some folks on the team think there are “better” ways to do this.

**Comment by nmav** [comment]
I'm certainly for any better way, what are they? I didn't notice any at the previous discussions in the mailing list.

**Comment by richsalz** [comment]
Nothing was offered beyond "we need a better way"

**Comment by mattcaswell** [comment]
Is this still relevant?

**Comment by richsalz** [comment]
Yes, it is a feature that some (including me) want to see.

**Comment by mattcaswell** [comment]
Ok. In which case in needs to be rebased.

**Comment by richsalz** [comment]
Sure.  More importantly, the concept needs review/discussion as an issue first.

**Comment by nmav** [comment]
The issue opened for it is #5041.
I'm closing as this was superseded by https://github.com/openssl/openssl/pull/4848

**Pull request closed without merging**